src/ngraph_mark_for_clustering.cc

```diff
@@ -266,20 +230,7 @@ Status MarkForClustering(Graph* graph) {
       confirmation_functions["AddN"] = always;
       confirmation_functions["AvgPool"] = always;
       confirmation_functions["AvgPoolGrad"] = [](Node* n, bool* result) {
-        Node* tf_orig_input_shape;
-        TF_RETURN_IF_ERROR(n->input_node(0, &tf_orig_input_shape));
-
-        std::vector<int64> tf_orig_input_shape_vec;
-        if (ExtractConstantData(tf_orig_input_shape,
-                                &tf_orig_input_shape_vec) !=
-                Status::OK() ||
-            tf_orig_input_shape_vec.size() != 4) {
-          *result = false;
-          return Status::OK();
-        }
-
-        n->AddAttr("_ngraph_avgpoolgrad_static_input_shape",
-                   tf_orig_input_shape_vec);
+        SetStaticInputs(n, {0});
```

Comment:
It seems that the body of the static_input nodes are almost same as the always function:
How about this:

ConfirmationFunction CreateConfirmationFunction(vector<int> staticinps = vector<int>()){
    auto cf = [staticinps](Node* n, bool* result) {
      auto make_positive = [n](int x){return  (x>0? x : n->num_inputs() - x);};
      *result = true;
      if (staticinps.size()>0){
        vector<int> pos_idx(staticinps.size());
        for(auto idx : staticinps)
          pos_idx.push_back(make_positive(idx));
        SetStaticInputs(n, staticinps);
      }
      return Status::OK();
    };
    return cf;
}

auto always = CreateConfirmationFunction();

and then we can have:
confirmation_functions["AvgPool"] = CreateConfirmationFunction({0});
or even negative numbers:
confirmation_functions["ConcatV2"] = CreateConfirmationFunction({-1});
and for normal cases:
confirmation_functions["BiasAdd"] = always; or confirmation_functions["BiasAdd"] = CreateConfirmationFunction();

Reply:
i likes it